pdata/internal/cmd/pdatagen/internal/base_slices.go

```diff
@@ -98,32 +99,57 @@ func Test${structName}_RemoveIf(t *testing.T) {
 }`
 
 const slicePtrTemplate = `// ${structName} logically represents a slice of ${elementName}.
-//
-// This is a reference type. If passed by value and callee modifies it, the
-// caller will see the modification.
-//
-// Must use New${structName} function to create new instances.
-// Important: zero-initialized instance is not valid for use.
-type ${structName} struct {
+type ${structName} interface {
+	common${structName}
+	At(ix int) ${elementName}
+}
+
+type Mutable${structName} interface {
+	common${structName}
+	At(ix int) Mutable${elementName}
+	EnsureCapacity(newCap int)
+	AppendEmpty() Mutable${elementName}
+	Sort(less func(a, b Mutable${elementName}) bool)
+}
+
+type common${structName} interface {
+	Len() int
+	CopyTo(dest Mutable${structName})
+	getOrig() *[]*${originName}
```

Comment:
> We still need an unexported func to disallow external implementations.

Why? I thought this should be able to play that role. Another reason for this is to simplify the access to the orig from interface instances passed as arguments, so instead of

```
func (es immutableSummaryDataPointSlice) CopyTo(dest MutableSummaryDataPointSlice) {
	destCap := cap(*dest.(mutableSummaryDataPointSlice).getOrig())
```

we have 

```
func (es immutableSummaryDataPointSlice) CopyTo(dest MutableSummaryDataPointSlice) {
	destCap := cap(*dest.getOrig())
```